### ISSUE TITLE
MPI process on  LocalCluster

### ISSUE THREAD

**Issue by srmnitc** [E-CAM/jobqueue_features#39]
Hi! I have a small question on the usage - Is it possible to run an MPI process on local cluster? The CustomSlurm cluster has all the necessary options, but would this work in a similar way for a local cluster?

**Comment by AdamWlodarczyk** [comment]
Hi, this is a really good question. For simple task  I would say that that using `mpi_task` and `on_cluster` (with registered `LocalCluster`) decorators combination would do that, but since you involve `CustomSLURMCluster` I would go for multiple iheritance with `CustomClusterMixin` and `LocalCluster`. Something like:
```
class CustomLocalCluster(CustomClusterMixin, LocalCluster):
     ...
```
I will check executability of above and let you know if you're still interested in this. I bet that would be a nice option for this.

**Comment by ocaisa** [comment]
If you want to run mpi executables then yes, and there's an example at https://github.com/E-CAM/jobqueue_features/blob/master/examples/mpi_tasks_mpiexec.py

However, if you would like to use something with python bindings and leverages is MPI then no. The problem in this case would be that nothing is ever started within an MPI runtime (by which I mean there is never any call to `mpirun` or similar) so the environment cannot get initialised. It would not be impossible to implement, probably along the lines of what @AdamWlodarczyk mentions above since we already have all the machinery, but it would take a bit of thought.

Which one of the two use cases are you considering?

**Comment by srmnitc** [comment]
The simple example would not suffice I believe. I will explain the scenario a bit more. I want to use the lammps python interface on multiple cores. You have an example with pylammps for this, but the difference is that I need to be able to gather values at runtime. My experience with MPI is super limited, so it might not even be possible, but could we for example just run the lammps `run` statement on the cluster.

To sum it up, create a lammps python object which can run on multiple cores, set up the run, then just get the [lammps run](https://lammps.sandia.gov/doc/run.html) on the dask cluster. I am working on a library for this purpose - you can see [here](https://github.com/pyiron/pylammpsmpi) and I thought it might be useful to integrate with dask. Is this possible?

**Comment by srmnitc** [comment]
One more thing - I would like the above scenario to work on any cluster - I was just asking about local cluster to think about feasilbility

**Comment by ocaisa** [comment]
So the linked code does a lot of similar stuff to what we try to do here, but the main difference is that you are maintaining a stateful instance of LAMMPS in the MPI job. That's not something we can do right now but that is possible with Dask actors (see the [example of Willi from our workshop last year](https://gist.github.com/willirath/10072950d4dd5083dc0dd1587f12feb4)).

One thing I don't quite understand is what your wrapper gets you? The LAMMPS python interface can already be executed in parallel. In the specific case of LAMMPS and our library, you can cheat pretty easily to get yourself a stateful instance. Use setters and getters to create a global LAMMPS instance that can survive outside of individual tasks like we do now for the MPI communicator (see https://github.com/E-CAM/jobqueue_features/blob/master/jobqueue_features/mpi_wrapper.py#L17 and how we manipulate it).

However, I am not suggesting you use our library. For the specific use case of the LAMMPS python interface, our library does not currently work as intended (see #18). The reason is that the LAMMPS bindings do not release the GIL, meaning Dask cannot send a heartbeat and so our workers appear to die. Your approach sidesteps the GIL problem and, for your use case, I think your library plus Dask actors (and Dask jobqueue) would be the way to go. We will have another workshop on Dask sometime this year, if you manage to get things working it would be great to have you!

On a side note, your library currently assumes that `mpiexec -n <>` will work anywhere, despite it being the standard this is probably not a wise assumption. I'd suggest you have `mpi_launcher="mpiexec"` and `mpi_launcher_args="-n "` to allow people to drop in more suitable replacements.

**Comment by srmnitc** [comment]
@ocaisa Thanks a lot for the feedback, the library was put of a [larger project](https://github.com/pyiron), I jumped in mainly  because the library was extremely convenient in accessing the state variables and so on. The way forward is to possibly have a simulation set up in a notebook and possibly run parts of it on a queuing system but have a continuous output that can be monitored. Our last workshop came to mind and that is how I checked this out. I will check out dask actors, and I will update you if I have things working. I am of course interested in dask related workshops.
Thanks for the `mpiexec` tip, I will check it out. I will close this issue for now, thanks for the quick feedback.

**Comment by srmnitc** [comment]
Hi @ocaisa - for the above discussion, dask actors were the right way to go. I was able to get it to work as expected. There is a small example [here](https://github.com/pyiron/pylammpsmpi/blob/master/notebooks/lammps_slurm_cluster.ipynb). Thanks for the suggestion!

**Comment by ocaisa** [comment]
That's good news @srmnitc! I think you have an interesting use case example for *interactive supercomputing*, there's a workshop at ISC20 that addresses that: https://www.interactivehpc.com/

You could submit an abstract there and see if they are interested (deadline is the end of the month)...and you would get a publication from it.

**NB**: ISC20 has actually been cancelled due to COVID-19 and will be online instead, for workshops like this it's still unclear what will happen. In the case of normal papers to ISC, they are still going to publish proceedings, so it might also be the case for workshops
